### PR TITLE
Allow `is_retained_metrics_relation` for indexes; turn on indexing for utilization metrics

### PR DESCRIPTION
**Pull request by umanwizard**
We want to turn on indexes for the queries used by the console to populate CPU/memory data, in order to decrease the latency of those queries as well as the load on S3. To do so, we need to allow indexes to be retained according to the metrics retention window, the same way tables and sources already can be.

This PR is a relatively straightforward port of the `is_retained_metrics_relation` logic to indexes. 

For testing, we modify the existing `test_utilization_hold` to run its queries in `mz_introspection`, verifying both that the indexes are available and that their since is properly held back.

Before landing, I will run this in staging for a day or two to make sure the memory usage is not too extreme.

### CODE REVIEW

File: src/adapter/src/catalog.rs
Comment (teskje) [inline src/adapter/src/catalog.rs:1685]: Nit: It feels a bit strange to call indexes "relations". Maybe that's just me.

**Review by teskje — COMMENTED** [review]
LGTM, pending Joe's approval.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:1685]: We specifically do no consider indexes to be relations in the code: https://github.com/MaterializeInc/materialize/blob/a0714eb8c26268dbd9ed6b0f49d8ae19dc573842/src/adapter/src/catalog.rs#L2109-L2124

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [inline src/environmentd/tests/sql.rs:1303]: I think that this no longer tests that the tables retention window works, am I reading that right? Maybe we should execute the query twice, once in the `mz_introspection` cluster using the index and once out of that cluster using the table.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:1685]: Either way, this matches the fields for tables, and I vaguely remember asking the same question then so it's probably fine.

File: src/adapter/src/coord/read_policy.rs
Comment (jkosh44) [inline src/adapter/src/coord/read_policy.rs:331]: Any specific reason to have this here instead of at the top of the file?

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/adapter/src/catalog.rs
Comment (umanwizard) [reply src/adapter/src/catalog.rs:1685]: I don't think there is a difference today, I just copied how it works for tables/sources.

Logically the difference might be that if `is_retained_metrics_relation` is `false` but `custom_logical_compaction_window` is set, it means the user has manually configured a retention window for an index of their choice, which we don't support at all yet so it's a moot point.

We can probably re-think the structure of this once we turn on custom introspection windows.

File: src/adapter/src/catalog.rs
Comment (umanwizard) [reply src/adapter/src/catalog.rs:1966]: That is a good point, I'll rename to `object`.

File: src/adapter/src/coord/ddl.rs
Comment (teskje) [reply src/adapter/src/coord/ddl.rs:675]: Okay, I see now that it differs on `entry.item().cluster_id()` `None/Some` values. But that doesn't really partition by storage/compute objects because sources can have clusters associated (and MVs are also storage objects, though I'm not sure if that matters here).

Comment (umanwizard) [reply src/adapter/src/coord/ddl.rs:675]: That's a good point, I suppose the logic is not fully correct for sources.

Perhaps I need to add a `fn is_compute_object(&self) -> bool` to `CatalogItem` which corresponds to which system the policies need to be set on.

**Pull request merged**
